rust/src/dataset.rs

```diff
@@ -480,6 +482,87 @@ impl Dataset {
         })
     }
 
+    /// Merge this dataset with another arrow Table / Dataset, and returns a new version of dataset.
+    ///
+    /// Parameters:
+    ///
+    /// - `stream`: the stream of [`RecordBatch`] to merge.
+    /// - `left_on`: the column name to join on the left side (self).
+    /// - `right_on`: the column name to join on the right side (stream).
+    ///
+    /// Returns: a new version of dataset.
+    ///
+    /// It performs a left-join on the two datasets.
+    pub async fn merge(
+        &mut self,
+        stream: &mut Box<dyn RecordBatchReader>,
+        left_on: &str,
+        right_on: &str,
+    ) -> Result<()> {
+        // Sanity check.
+        if self.schema().field(left_on).is_none() {
+            return Err(Error::invalid_input(format!(
+                "Column {} does not exist in the left side dataset",
+                left_on
+            )));
+        };
+        let right_schema = stream.schema();
+        if right_schema.field_with_name(right_on).is_err() {
+            return Err(Error::invalid_input(format!(
+                "Column {} does not exist in the right side dataset",
+                right_on
+            )));
+        };
+        for field in right_schema.fields() {
+            if field.name() == right_on {
+                continue;
+            }
+            if self.schema().field(field.name()).is_some() {
+                return Err(Error::invalid_input(format!(
+                    "Column {} exists in both sides of the dataset",
+                    field.name()
+                )));
+            }
+        }
+
+        // Hash join
+        let joiner = Arc::new(HashJoiner::try_new(stream, right_on).await?);
+        let new_schema: Schema = self.schema().merge(joiner.out_schema().as_ref())?;
```

Comment:
is the `left_on` and `right_on` columns de-duplicated here?

Reply:
`joiner.out_schema()` returns the schema of the RHS with the `right_on` column removed. So right now we always drop the right_on column. Is that alright? Or do we sometimes want to keep both `left_on` and `right_on`?

Reply:
Oh, we want to drop `right_on` column.  Have not read this code for a while, wanted to refresh my understanding there.

Reply:
Sound good. Confirmed this is the old behavior

https://github.com/lancedb/lance/blob/78e5af7dec1a9602823797ba18eafc632ba727c3/cpp/src/lance/arrow/hash_merger.cc#L159-L162